### ISSUE TITLE
[BUG] Syntax error in C array declaration and initialization

### ISSUE THREAD

**Issue by GalaxySnail** [cython/cython#5179]
### Describe the bug

I found it in #5177.

``` cython
def main():
    cdef int arr[4] = [1, 2, 3, 4]
```

This code doesn't compile:

```
Error compiling Cython file:
------------------------------------------------------------
...
def main():
    cdef int arr[4] = [1, 2, 3, 4]
                    ^
------------------------------------------------------------

test.pyx:2:20: Syntax error in C variable declaration
```

But these two do:

``` cython
def main():
    cdef int arr1[4]
    arr1 = [1, 2, 3, 4]

    cdef int[4] arr2 = [1, 2, 3, 4]
```

I think it is already known, because it has been mentioned in the documentation:

- https://github.com/cython/cython/blob/b2cad768499cf5763251fe757491e9fc24a1f583/docs/src/userguide/language_basics.rst#L124-L129

but I couldn't find the related issue.

Anyway, if it can't be fixed easily, it will be good to document it more clearly.

### Environment

Cython version: both `0.29.32` and `3.0.0a11`

**Comment by matusvalo** [comment]
Just a thought. Should Cython support both syntaxes to declare array?
```cython
cdef int arr1[4]   # C style array declaration
cdef int[4] arr1   # Java style array declaration
```

it seems that both of them generate same C code but the Java style (`cdef int[4] arr1`) is more consistent because:

1. Memoryviews are defined in the same way: `cdef int [:, :, :] narr_view = narr` [1]
2. As mentioned in the bug report Cython currently supports `cdef int[4] arr2 = [1, 2, 3, 4]` but not `cdef int arr[4] = [1, 2, 3, 4]`
3. Defining fused types uses syntax of `TYPE[n]`:
    ```cython
    ctypedef fused my_fused_type:
        int[3]
        float[2]
    ```

Hence my proposal is following. Does it make sense to soft-deprecated: use only Java style array definition (`cdef int[4] arr1`) in Cython Documentation in all places but still support the C style (to keep backward compatibility).


Any thoughts?

[1] https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html

**Comment by GalaxySnail** [comment]
> it seems that both of them generate same C code but the Java style (cdef int[4] arr1) is more consistent
>
> Soft-deprecate C style definition - use only Java style array definition (cdef int[4] arr1) in Cython Documentation in all places but still support the C style (to keep backward compatibility).

Fair enough.

BTW, Java style `cdef int[4] a, b` is different from C style `cdef int a[4], b` (whether `b` is an areay or an int), but the latter seems to have been deprecated for almost 10 years: https://github.com/cython/cython/commit/9e3a2d7ae9fce31bfbba19f39fc882e6ded78462.

However, the documentation said:

- https://github.com/cython/cython/blob/d888842ed8efec0d859dc5978900bb08161d7835/docs/src/userguide/language_basics.rst#L89-L91

which will generate warnings:

```
warning: test.pyx:3:19: Non-trivial type declarators in shared declaration (e.g. mix of pointers and values). Each pointer declaration should be on its own line.
warning: test.pyx:3:25: Non-trivial type declarators in shared declaration (e.g. mix of pointers and values). Each pointer declaration should be on its own line.
```

IMO it's better to mention it in the documentation.

**Comment by da-woods** [comment]
I'll leave this open for now. I think I'd accept a simple parser change to make the other syntax work. But I suspect it's low priority and may not happen

**Comment by williamhCode** [comment]
I was just wondering how would I declare an array of pointers with java style declaration. I stumbled across this while trying to fit some linked lists in an array.
```cython
cdef Node *data[4]
data = [NULL, NULL, NULL, NULL]
```
This is the only way that I can think of initialising data for now.  
I was expecting it to be something like this to work:
```cython
cdef Node*[4] data = [NULL, NULL, NULL, NULL]
```

**Comment by scoder** [comment]
This should work:

> cdef Node* data = [NULL, NULL, NULL, NULL]

**Comment by williamhCode** [comment]
Doing `cdef Node** data = [NULL, NULL, NULL, NULL]` works as a replacement thx. It's interesting how using `type* var_name` in Cython is like `type var_name[]` in C where u can omit the length if being initialized.

**Comment by GalaxySnail** [comment]
If I understand correctly, "soft-deprecating" something doesn't mean that we will actually deprecate it. "soft-deprecated" only means it's not recommended, but will be still supported without any warnings. So if you really need the C style syntax, you can use it as expected, it's OK.

However, I agree that:

``` cython
cdef Node*[4] data = [NULL, NULL, NULL, NULL]
```

can be a useful feature that will improve code consistency.

**Comment by matusvalo** [comment]
> "soft-deprecated" only means it's not recommended, but will be still supported without any warnings.

Yes. It should be the preferred way and should be used in all examples in documentation.
